libraries/push/impl/src/main/kotlin/io/element/android/libraries/push/impl/troubleshoot/NotificationTest.kt

```diff
@@ -0,0 +1,97 @@
+import io.element.android.libraries.core.notifications.NotificationTroubleshootTest
+import io.element.android.libraries.core.notifications.NotificationTroubleshootTestDelegate
+import io.element.android.libraries.core.notifications.NotificationTroubleshootTestState
+import io.element.android.libraries.di.AppScope
+import io.element.android.libraries.push.impl.R
+import io.element.android.libraries.push.impl.notifications.NotificationDisplayer
+import io.element.android.libraries.push.impl.notifications.factories.NotificationCreator
+import io.element.android.services.toolbox.api.strings.StringProvider
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withTimeoutOrNull
+import timber.log.Timber
+import javax.inject.Inject
+import kotlin.time.Duration.Companion.seconds
+
+@ContributesMultibinding(AppScope::class)
+class NotificationTest @Inject constructor(
+    private val notificationCreator: NotificationCreator,
+    private val notificationDisplayer: NotificationDisplayer,
+    private val notificationClickHandler: NotificationClickHandler,
+    private val stringProvider: StringProvider,
+) : NotificationTroubleshootTest {
+    override val order = 50
+    private val delegate = NotificationTroubleshootTestDelegate(
+        defaultName = stringProvider.getString(R.string.troubleshoot_notifications_test_display_notification_title),
+        defaultDescription = stringProvider.getString(R.string.troubleshoot_notifications_test_display_notification_description),
+        fakeDelay = NotificationTroubleshootTestDelegate.SHORT_DELAY,
+    )
+    override val state: StateFlow<NotificationTroubleshootTestState> = delegate.state
+
+    override suspend fun run(coroutineScope: CoroutineScope) {
+        delegate.start()
+        val notification = notificationCreator.createDiagnosticNotification()
+        val result = notificationDisplayer.displayDiagnosticNotification(notification)
+        if (result) {
+            coroutineScope.listenToNotificationClick()
+            delegate.updateState(
+                description = stringProvider.getString(R.string.troubleshoot_notifications_test_display_notification_waiting),
+                status = NotificationTroubleshootTestState.Status.WaitingForUser
+            )
+        } else {
+            delegate.updateState(
+                description = stringProvider.getString(R.string.troubleshoot_notifications_test_display_notification_permission_failure),
+                status = NotificationTroubleshootTestState.Status.Failure(false)
+            )
+        }
+    }
+
+    private fun CoroutineScope.listenToNotificationClick() = launch {
+        val job = launch {
+            notificationClickHandler.state.first()
+            Timber.d("Notification clicked!")
+        }
+        val s = withTimeoutOrNull(30.seconds) {
```

Comment:
This seems a bit hard to understand, IMO. Maybe we could use just `withTimeout` inside a `try/catch` block or `runCatching`, the successful path leading to the `Success` state and the catch clause for `TimeoutCancellationException` emitting the `Failure` case?